src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs

```diff
@@ -14,8 +14,38 @@ fn in_match() {
     let foo = Foo::Bar { field: &y };
     //~^ ERROR `y` does not live long enough
     match foo {
-        Foo::Bar::<'static> { field: _z } => {
-        }
+        Foo::Bar::<'static> { field: _z } => {}
+    }
+}
+
+fn in_let_2() {
+    let y = 22;
+    let foo = Foo::Bar { field: &y };
+    //~^ ERROR `y` does not live long enough
+    let Foo::<'static>::Bar { field: _z } = foo;
+}
+
+fn in_match_2() {
+    let y = 22;
+    let foo = Foo::Bar { field: &y };
+    //~^ ERROR `y` does not live long enough
+    match foo {
+        Foo::<'static>::Bar { field: _z } => {}
+    }
+}
+
+// FIXME: the following seem to be incorrect and would be a regression.
+fn in_let_3() {
+    let y = 22;
+    let foo = Foo::Bar { field: &y };
+    let Foo::Bar { field: _z } = foo;
+}
+
+fn in_match_3() {
+    let y = 22;
+    let foo = Foo::Bar { field: &y };
+    match foo {
+        Foo::Bar { field: _z } => {}
```

Comment:
This should be passing and it isn't, I'm guessing I need to be using `Value` where I'm using `Type` somewhere?

Reply:
cc @matthewjasper

Reply:
CC @matthewjasper: @eddyb mentioned that the "user type annotation" system for MIR borrowck might be needing adjustment for these changes. Could you point me in the right direction?

Reply:
```
[DEBUG rustc_mir::borrow_check::type_check] check_user_type_annotations: user_type_annotations=[CanonicalUserTypeAnnotation { user_ty: Canonical { max_universe: U0, variables: [], value: TypeOf(DefId(0:3 ~ pattern_substs_on_brace_enum_variant[317d]::Foo[0]), UserSubsts { substs: [ReStatic], user_self_ty: None }) }, span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:40:15: 40:37, inferred_ty: Foo<'_> }, CanonicalUserTypeAnnotation { user_ty: Canonical { max_universe: U0, variables: [], value: TypeOf(DefId(0:3 ~ pattern_substs_on_brace_enum_variant[317d]::Foo[0]), UserSubsts { substs: [ReStatic], user_self_ty: None }) }, span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:41:9: 41:31, inferred_ty: Foo<'_> }, CanonicalUserTypeAnnotation { user_ty: Canonical { max_universe: U0, variables: [], value: TypeOf(DefId(0:3 ~ pattern_substs_on_brace_enum_variant[317d]::Foo[0]), UserSubsts { substs: [ReStatic], user_self_ty: None }) }, span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:41:9: 41:31, inferred_ty: Foo<'_> }]
```

Reply:
```
[DEBUG rustc_mir::borrow_check::renumber] visit_region: region='_#2r
[DEBUG rustc_mir::borrow_check::renumber] visit_region(region=ReErased, location=bb0[7])
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=ReErased)
[DEBUG rustc_mir::borrow_check::renumber] visit_region: region='_#3r
[DEBUG rustc_mir::borrow_check::renumber] visit_substs(substs=[ReStatic], location=bb0[8])
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=[ReStatic])
[DEBUG rustc_mir::borrow_check::renumber] visit_substs: substs=['_#4r]
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=&'static u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=(), ty_context=ReturnTy(SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:38:1: 42:2, scope: scope[0] }))
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=())
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=()
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=(), ty_context=LocalDecl { local: _0, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:38:15: 38:15, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=())
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=()
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=u32, ty_context=LocalDecl { local: _1, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:39:9: 39:10, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=u32
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=Foo<'static>, ty_context=LocalDecl { local: _2, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:40:9: 40:12, scope: scope[1] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=Foo<'static>)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=Foo<'_>
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=&'static u32, ty_context=LocalDecl { local: _3, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:40:33: 40:35, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=&'static u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=&u32
```

vs before the change

```
[DEBUG rustc_mir::borrow_check::renumber] visit_region: region='_#2r
[DEBUG rustc_mir::borrow_check::renumber] visit_region(region=ReErased, location=bb0[7])
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=ReErased)
[DEBUG rustc_mir::borrow_check::renumber] visit_region: region='_#3r
[DEBUG rustc_mir::borrow_check::renumber] visit_substs(substs=[ReScope(Remainder { block: 26, first_statement_index: 1})], location=bb0[8])
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=[ReScope(Remainder { block: 26, first_statement_index: 1})])
[DEBUG rustc_mir::borrow_check::renumber] visit_substs: substs=['_#4r]
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=&u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=(), ty_context=ReturnTy(SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:5:1: 10:2, scope: scope[0] }))
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=())
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=()
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=(), ty_context=LocalDecl { local: _0, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:5:13: 5:13, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=())
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=()
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=u32, ty_context=LocalDecl { local: _1, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:6:9: 6:10, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=u32
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=Foo<'_>, ty_context=LocalDecl { local: _2, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:7:9: 7:12, scope: scope[1] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=Foo<'_>)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=Foo<'_>
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=&u32, ty_context=LocalDecl { local: _3, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:7:33: 7:35, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=&u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=&u32
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=&u32, ty_context=LocalDecl { local: _4, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:7:33: 7:35, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=&u32)
```